File: HINT-python/hintController.py
# -*- coding: utf-8 -*-
"""
Created on Thu May  5 13:52:26 2022

@author: Alexander Mueller
"""

import numpy as np;
import json;
import soundfile as sf;
import sounddevice as sd;
import csv;
import datetime as dt;
import os;
import hintUtilities as util

#%%

    
# %%
hintDir = 'german-hint-adaptive-48kHz\\';
base_type = 'wav';
importDir = 'G:\VRHINT-misc\HINT-python\german-hint-adaptive-48kHz\\';

# Test setup
# lists open for test
availableTestLists = 10;
# number of lists used in the test
numTestLists = 5;
# number of sentences in each list
listSentences = 20;

# %%
# min/max available SNR
minSNR = -16;
maxSNR = 2;

# practice setup
practiceList = 12;
practiceRounds = 5;
practiceCondition = "noiseLeft";

# %% Type definitions etc

# this format works for sentence indices, SNR values and hit Quotes
templateData = np.zeros(listSentences);
templateCondition = "noiseLeft";
templateListIndex = 12;

resultTemplate = {
    "ListIndex": templateListIndex,
    "Condition": templateCondition,
    "SNRs": templateData,
    "HitQuotes:": templateData,
    "Time": dt.datetime.now().strftime("%d-%m-%y--%H-%M-%S")};

# allocate numTestLists structs to store results
resultStorage = [resultTemplate for k in range(numTestLists)];
    

with open('lqConditions.csv', mode ='r')as file:
  lqConditions = list(csv.reader(file));
        
with open('lqLists.csv', mode ='r')as file:
  lqLists = list(csv.reader(file));

# convert List<List<str>> to List<List<int>>
for i in range(len(lqLists)):
    lqLists[i] = [int(j) for j in lqLists[i]];


#%% user login
print("Enter participants name:");
name = input();
# add system to check if username has already been used!
# create file name based on user name
resultFileName = "results\\results-%s-%s.json" % (name,  dt.datetime.now().strftime("%d-%m-%y--%H-%M-%S"));
practiceFileName = "results\\results-%s-%s.json" % (name,  dt.datetime.now().strftime("%d-%m-%y--%H-%M-%S"));

#%% load noise
calibrationNoise, fs = sf.read(importDir + "NBNoise1000.wav");
noise, fs = sf.read(importDir + "noiseGR_male.wav");
sd.default.samplerate = fs;

#%% Initialize playrec
#init_playrec(fs);

# define channel map
#ChMap= [1 3 5]; % uni setup
#ChLeft = 1;
#ChFront = 3;
#ChRight = 5;

ChMap = [1, 2]; # at home
ChLeft = 1;
ChRight = 2;
# only have 2 channels here so do this for testing...
ChFront = 2;

#%% Load counterbalanced test order
userIndex = len(os.listdir('results'));
print("User Index is: " + str(userIndex));
print("Is this correct? y/n");
if str(input()) != 'y':
    exit();

# this will be determined by number of result files!
testLists = lqLists[userIndex % availableTestLists];

# pre-allocate text array
testConditions = ["emptyCondition" for k in range(numTestLists)]; 

# userIndex determines start line of lqConditions matrix
# 5th condition is taken from the next line (considering wrapping!)
testConditions[0:4] = lqConditions[userIndex % len(lqConditions)];
testConditions[4] = lqConditions[(userIndex + 1) % len(lqConditions)][0];

#%% start practice condition 
sentences = util.loadListSentences(practiceList, hintDir);
randOrder = np.random.permutation(range(20));

# store SNR for each sentence
listSNR = np.zeros(20);
# store hitQuote for each sentence
hitQuotes = np.zeros(20);

# starting at 0 dB
currentSNR = 0;
noiseIndex = 1;

for i in range(practiceRounds):
    # get random index
    index = randOrder[i];
    # load current sentence
    sentencePath = [hintDir + '0' + str(practiceList) + '\-0dB\Ger_male00' + str(index) + '.wav'];

    print(["Current playback level: " + str(currentSNR)]);
    print(["Round " + str(i) + " out of " + str(practiceRounds)]);
    # audio files are labeled from Ger_male001 and not Ger_male000 so add '1'
    curr_sent = util.loadSentenceAudio(practiceList, index + 1, currentSNR, importDir);
        
    #noiseSegment = noise[0:len(curr_sent)];
    #[noiseSegment, noiseIndex] = circularNoise(noise, sentLen, noiseIndex);

    audioBuffer = np.array([curr_sent, noise[0:len(curr_sent)]]).transpose();
    
    # get length of current sentence
    sentenceLength = len(sentences[index].split());
   
    # print out current sentence string
    print("Sentence length: " + str(sentenceLength))
    print(sentences[index]);
    
    sentLen = len(curr_sent);

    if practiceCondition == "noiseLeft":
        sd.play(audioBuffer, blocking = 'true', mapping = [ChFront, ChLeft]);
    elif practiceCondition == "noiseRight":
        sd.play(audioBuffer, blocking = 'true', mapping = [ChFront, ChRight]);
    elif practiceCondition == "noiseFront":
        sd.play(curr_sent + noise[0:len(curr_sent)], blocking = 'true', mapping = [ChFront]);
    else:
        sd.play(audioBuffer[0], blocking = 'true', mapping = [ChFront]);


    # get experimenter feedback
    print("How many words have been correct?");
    correctWords = int(input());
    #print(type(correctWords));

    # alternative: just ask if below or above 50% hit
    # less data but quicker
    while correctWords > sentenceLength:
        print("Invalid input. Try again!");
        print("How many words have been correct?");
        correctWords = input();
    
    print('Sentence len: ' + str(sentenceLength) + ' correct: ' + str(correctWords));
    hitQuote = correctWords / sentenceLength;
    print("HitQuote: " + str(hitQuote));


    # adapt SNR based on hitQuote
    if i < 4 :        
        if hitQuote < 0.5:
            currentSNR = currentSNR + 4;
        else:
            currentSNR = currentSNR - 4;
    else:
        # store sentence data
        hitQuotes[i] = hitQuote;
        listSNR[i] = currentSNR;
        
        # adapt SNR based on hitQuote
        if hitQuote < 0.5:
            currentSNR = currentSNR + 2;
        else:
            currentSNR = currentSNR - 2;


    # SNR sanity check
    if currentSNR < minSNR:
        currentSNR = minSNR;
        print(["Warning: reached min SNR!" + str(minSNR)]);
    elif currentSNR > maxSNR:
        currentSNR = maxSNR;
        print(["Warning: reached max SNR!" + str(maxSNR)]);


print("Practice done!");

#%% Test JSON format
# fill struct with list data
pracRes = resultTemplate;
pracRes["ListIndex"] = practiceList;
pracRes["Condition"] = practiceCondition;
pracRes["SNRs"] = listSNR;
pracRes["HitQuotes"] = hitQuotes;

# parse struct into json
practiceResults = json.dumps(pracRes, indent = 4, cls=util.NumpyEncoder);
# store json into file
with open(practiceFileName, "w") as outfile:
    outfile.write(practiceResults)

#%% Test procedure
for j in range(numTestLists):

    sentences = util.loadListSentences(testLists[j], hintDir);
    randOrder = np.random.permutation(range(20));
    currentCondition = testConditions[j];
    
    # store SNR for each sentence
    listSNR = np.zeros(20);
    # store hitQuote for each sentence
    hitQuotes = np.zeros(20);
    
    # starting at 0 dB
    currentSNR = 0;
    noiseIndex = 1;


    print(["Starting List " + str(testLists[j]) + " with condition" + testConditions[j]]);

    for i in range(listSentences):
        # get random index
        index = randOrder[i];
        # load current sentence
        sentencePath = [hintDir +'0' + str(testLists[j]) + '\-0dB\Ger_male00' + str(index) + '.wav'];
    
        print(["Current playback level: " + str(currentSNR)]);
        print(["Round " + str(i) + " out of " + str(listSentences)]);
        # audio files are labeled from Ger_male001 and not Ger_male000 so add '1'
        curr_sent = util.loadSentenceAudio(testLists[j], index + 1, currentSNR, hintDir);
    
        sentLen = len(curr_sent);
        #[noiseSegment, noiseIndex] = circularNoise(noise, sentLen, noiseIndex);

        if testConditions[j] == "noiseLeft":
            sd.play(audioBuffer, blocking = 'true', mapping = [ChFront, ChLeft]);
        elif testConditions[j] == "noiseRight":
            sd.play(audioBuffer, blocking = 'true', mapping = [ChFront, ChRight]);
        elif testConditions[j] == "noiseFront":
            sd.play(curr_sent + noise[0:len(curr_sent)], blocking = 'true', mapping = [ChFront]);
        else:
            sd.play(audioBuffer[0], blocking = 'true', mapping = [ChFront]);
            

        # get length of current sentence
        sentenceLength = len(sentences[index].split());

        # print out current sentence string
        print("Sentence length: " + str(sentenceLength))
        print(sentences[index]);    

        # get experimenter feedback
        print("How many words have been correct?");
        correctWords = input();
    
        # alternative: just ask if below or above 50% hit
        # less data but quicker
        while correctWords > sentenceLength:
            print("Invalid input. Try again!");
            print("How many words have been correct?");
            correctWords = input();
        
        print(['Sentence len: ' + str(sentenceLength) + ' correct: ' + str(correctWords)]);
        hitQuote = correctWords / sentenceLength;
        print("HitQuote: " + str(hitQuote));
        
        # adapt SNR based on hitQuote
        if i < 4:
            if hitQuote < 0.5:
                currentSNR = currentSNR + 4;
            else:
                currentSNR = currentSNR - 4;
        else:
            
            # store sentence data
            hitQuotes[i] = hitQuote;
            listSNR[i] = currentSNR;
            
            # adapt SNR based on hitQuote
            if hitQuote < 0.5:
                currentSNR = currentSNR + 2;
            else:
                currentSNR = currentSNR - 2;

    
        # SNR sanity check
        if currentSNR < minSNR:
            currentSNR = minSNR;
            print(["Warning: reached min SNR!" + str(minSNR)]);
        elif currentSNR > maxSNR:
            currentSNR = maxSNR;
            print(["Warning: reached max SNR!" + str(maxSNR)]);
        
    

    print(["List" + str(j) + " out of" + str(numTestLists) + " done!"]);

    # store data in resultStorage
    resultStorage[j]["ListIndex"] = testLists[j];
    resultStorage[j]["Condition"] = testConditions[j];
    resultStorage[j]["ListSNRs"] = listSNR;
    resultStorage[j]["ListHitQuotes"] = hitQuotes;

   
#%% Write results into file
# parse dict into json
results = json.dumps(resultStorage, indent = 4, cls=util.NumpyEncoder);
# store json into file
with open(resultFileName, "w") as outfile:
    outfile.write(results)

File: HINT-matlab/hintController.m

%% Path stuff

addpath('Functions')  
addpath('german-hint-adaptive-48kHz/')  
addpath('Third-party')  
%savepath matlab/myfiles/pathdef.m

hintDir = 'german-hint-adaptive-48kHz\';

%% Test setup
% lists open for test
availableTestLists = 10;
% number of lists used in the test
numTestLists = 5;
% number of sentences in each list
listSentences = 20;

% min/max available SNR
minSNR = -16;
maxSNR = 2;

% practice setup
practiceList = 12;
practiceRounds = 5;
practiceCondition = "noiseFront";

%% Type definitions etc
tempAudioStruct = struct('AudioData', 1, 'Channel', 1);

% this format works for sentence indices, SNR values and hit Quotes
templateData = zeros(listSentences, 1);
templateCondition = "noiseLeft";
templateListIndex = 12;

resTemplate = struct('ListIndex', templateListIndex, 'Condition', templateCondition, 'ListSNRs', templateData, 'ListHitQuotes', templateData);

% allocate numTestLists structs to store results
resultStorage = repmat(resTemplate, numTestLists, 1 ); 

%% user login
prompt = "Enter participants name:\n";
name = input(prompt, "s");

%% pre-allocate struct array for results

% add system to check if username has already been used!
% create file name based on user name
%jsonFileName = sprintf("results-%s-%s.json", name, datestr(now, 'dd-mm-yyyy-hh:MM'));
jsonFileName = sprintf("results-%s.json", name);

%% load stimuli
% load calibration noise
calibrationNoise = audioread([hintDir 'NBNoise1000.wav']);
[noise,fs] = audioread([hintDir 'noiseGR_male.wav']);

%% Initialize playrec
init_playrec(fs);

% define channel map
%ChMap= [1 3 5]; % uni setup
%ChLeft = 1;
%ChFront = 3;
%ChRight = 5;

ChMap = [1 2]; % at home
ChLeft = 1;
ChRight = 2;
% only have 2 channels here so do this for testing...
ChFront = 2;
save('setup.mat');

%% Load counterbalanced test order

% get sentence list order (5 lists total)
testLists = randperm(availableTestLists, numTestLists);

% pre-allocate text array
testConditions = repmat("emptyCondition", numTestLists, 1 ); 

% assign random order of test conditions
% REPLACE THIS WITH LATIN SQUARES!
for i=1:numTestLists
    randNum = randi(4,1);

    switch(randNum)
        case 1
            testConditions(i) = "quiet";
        case 2
            testConditions(i) = "noiseFront";
        case 3
            testConditions(i) = "noiseLeft";
        case 4
            testConditions(i) = "noiseRight";
    end
end



%% start practice condition 
sentences = loadListSentences(practiceList, hintDir);
randOrder = randperm(20);

% store SNR for each sentence
listSNR = zeros(20, 1);
% store hitQuote for each sentence
hitQuotes = zeros(20, 1);

% starting at 0 dB
currentSNR = 0;
noiseIndex = 1;
audioStruct = repmat(tempAudioStruct, 2, 1); 

for i=1:practiceRounds
    % get random index
    index = randOrder(i);
    % load current sentence
    sentencePath = [hintDir '0' int2str(practiceList) '\-0dB\Ger_male00' num2str(index) '.wav'];

    disp(["Current playback level: " int2str(currentSNR)]);
    disp(["Round " int2str(i) " out of " int2str(practiceRounds)]);
    [curr_sent, fs] = loadSentenceAudio(practiceList, index, currentSNR, hintDir);

    sentLen = size(curr_sent, 1);
    [noiseSegment, noiseIndex] = circularNoise(noise, sentLen, noiseIndex);

    audioStruct(1).AudioData = curr_sent;
    audioStruct(1).Channel = ChFront;

    audioStruct(2).AudioData = noiseSegment;

    if practiceCondition == "noiseLeft"
        audioStruct(2).Channel = ChLeft;
    elseif practiceCondition == "noiseRight"
        audioStruct(2).Channel = ChRight;
    elseif practiceCondition == "noiseFront"
        audioStruct(2).Channel = ChFront;
    else
        audioStruct(2).AudioData = 0;
    end

    buffer = combineAudioFiles(audioStruct, sentLen);

    % play current sentence and noise
    playbackID = playrec('play', buffer, ChMap);

    % get length of current sentence
    sentenceLength = numel(strsplit(sentences(index)));

    % print out current sentence string
    disp([sentences(index) sentenceLength]);

    % only prompt this after playback is done
    while ~playrec('isFinished', playbackID) end

    % get experimenter feedback
    prompt = "How many words have been correct?";
    correctWords = input(prompt);
    disp(correctWords);

    % alternative: just ask if below or above 50% hit
    % less data but quicker
    while correctWords > sentenceLength
        disp("Invalid input. Try again!");
        prompt = "How many words have been correct?";
        correctWords = input(prompt);
    end
    disp(['Sentence len: ' int2str(sentenceLength) ' correct: ' int2str(correctWords)]);
    hitQuote = correctWords / sentenceLength;
    disp(hitQuote);

    % store sentence data
    hitQuotes(i) = hitQuote;
    listSNR(i) = currentSNR;

    % adapt SNR based on hitQuote
    if i < 4         
        if hitQuote < 0.5
            currentSNR = currentSNR + 4;
        else
            currentSNR = currentSNR - 4;
        end
    else
        % adapt SNR based on hitQuote
        if hitQuote < 0.5
            currentSNR = currentSNR + 2;
        else
            currentSNR = currentSNR - 2;
        end
    end

    % SNR sanity check
    if currentSNR < minSNR
        currentSNR = minSNR;
        disp(["Warning: reached min SNR!" int2str(minSNR)]);
    elseif currentSNR > maxSNR
        currentSNR = maxSNR;
        disp(["Warning: reached max SNR!" int2str(maxSNR)]);
    end

end

disp("Practice done!");

%% Test JSON format
% create file name based on user name
jsonFile = sprintf("practice-%s.json", name);
% fill struct with list data
s = struct('ListIndex', practiceList, 'Condition', practiceCondition, 'ListSNRs', listSNR, 'ListHitQuotes', hitQuotes);
% parse struct into json
practiceResults = jsonencode(s,PrettyPrint=true);
% store json into file
fid = fopen(jsonFile,'w');
fprintf(fid,'%s', practiceResults);
fclose(fid);


%% Store test data into json file
% parse struct into json
resultJSON = jsonencode(resultStorage,PrettyPrint=true);
% store json into file
fid = fopen(jsonFileName,'w');
fprintf(fid,'%s', resultJSON);
fclose(fid);


%% Test procedure
for j=1:numTestLists

    sentences = loadListSentences(testLists(j), hintDir);
    randOrder = randperm(20);
    currentCondition = testConditions(j);
    
    % store SNR for each sentence
    listSNR = zeros(20, 1);
    % store hitQuote for each sentence
    hitQuotes = zeros(20, 1);
    
    % starting at 0 dB
    currentSNR = 0;
    noiseIndex = 1;

    if testConditions(j) == "noiseLeft"
        audioStruct(2).Channel = ChLeft;
    elseif testConditions(j) == "noiseRight"
        audioStruct(2).Channel = ChRight;   
    elseif testConditions(j) == "noiseFront"
        audioStruct(2).Channel = ChFront;
    end

    disp(["Starting List " int2str(testLists(j)) " with condition" testConditions(j)]);

    for i=1:listSentences
        % get random index
        index = randOrder(i);
        % load current sentence
        sentencePath = [hintDir '0' int2str(testLists(j)) '\-0dB\Ger_male00' num2str(index) '.wav'];
    
        disp(["Current playback level: " int2str(currentSNR)]);
        disp(["Round " int2str(i) " out of " int2str(listSentences)]);
        [curr_sent, fs] = loadSentenceAudio(testLists(j), index, currentSNR, hintDir);
    
        sentLen = size(curr_sent, 1);
        [noiseSegment, noiseIndex] = circularNoise(noise, sentLen, noiseIndex);
    
        audioStruct(1).AudioData = curr_sent;
        audioStruct(1).Channel = ChFront;

        if testConditions(j) ~= "quiet"
            audioStruct(2).AudioData = noiseSegment;
        else
            audioStrcut(2).AudioData = 0;
        end

        buffer = combineAudioFiles(audioStruct, sentLen);
        % play current sentence
        playbackID = playrec('play', curr_sent, ChMap);
    

        % get length of current sentence
        sentenceLength = numel(strsplit(sentences(index)));

        % print out current sentence string
        disp([sentences(index) sentenceLength]);      
        
        % only prompt this after playback is done
        while ~playrec('isFinished', playbackID) end

        % get experimenter feedback
        prompt = "How many words have been correct?";
        correctWords = input(prompt);
        disp(correctWords);
    
        % alternative: just ask if below or above 50% hit
        % less data but quicker
        while correctWords > sentenceLength
            disp("Invalid input. Try again!");
            prompt = "How many words have been correct?";
            correctWords = input(prompt);
        end
        disp(['Sentence len: ' int2str(sentenceLength) ' correct: ' int2str(correctWords)]);
        hitQuote = correctWords / sentenceLength;
        disp(hitQuote);
    
        % store sentence data
        hitQuotes(i) = hitQuote;
        listSNR(i) = currentSNR;
    
        % adapt SNR based on hitQuote
        if i < 4         
            if hitQuote < 0.5
                currentSNR = currentSNR + 4;
            else
                currentSNR = currentSNR - 4;
            end
        else
            % adapt SNR based on hitQuote
            if hitQuote < 0.5
                currentSNR = currentSNR + 2;
            else
                currentSNR = currentSNR - 2;
            end
        end
    
        % SNR sanity check
        if currentSNR < minSNR
            currentSNR = minSNR;
            disp(["Warning: reached min SNR!" int2str(minSNR)]);
        elseif currentSNR > maxSNR
            currentSNR = maxSNR;
            disp(["Warning: reached max SNR!" int2str(maxSNR)]);
        end
    
    end

    disp(["List" int2str(j) " out of" int2str(numTestLists) " done!"]);

    % store data in resultStorage
    resultStorage(j).ListIndex = testLists(j);
    resultStorage(j).Condition = testConditions(j);
    resultStorage(j).ListSNRs = listSNR;
    resultStorage(j).ListHitQuotes = hitQuotes;
end  
   
%% Write results into file
% parse struct into json
results = jsonencode(resultStorage,PrettyPrint=true);
% store json into file
fid = fopen(jsonFile,'w');
fprintf(fid,'%s', results);
fclose(fid);

       


